File: CBTReframe/Services/GeminiService.swift
import Foundation

/// Google Gemini：`generateContent`（API Key 来自 Google AI Studio / Vertex）
struct GeminiService: AIServiceProtocol {
    let provider = AIProvider.gemini

    /// 与 `URLSession.shared` 分离：默认配置的 `timeoutIntervalForRequest` 仅 60s，慢模型可能在首包前被切断。
    private static let generateSession: URLSession = {
        let config = URLSessionConfiguration.ephemeral
        config.timeoutIntervalForRequest = 300
        config.timeoutIntervalForResource = 600
        return URLSession(configuration: config)
    }()

    func reframe(
        thought: String,
        mood: String,
        hasAkathisia: Bool,
        model: AIModel,
        mode: ReframeMode,
        style: ResponseStyle,
        template: PromptTemplate,
        strategy: ResponseStrategy
    ) async throws -> AnalysisResult {
        guard let apiKey = KeychainManager.shared.load(key: provider.rawValue),
              !apiKey.isEmpty else {
            throw AIServiceError.noAPIKey
        }

        let systemPrompt = PromptBuilder.buildSystemPrompt(
            mode: mode,
            style: style,
            template: template,
            strategy: strategy,
            mood: mood,
            hasAkathisia: hasAkathisia
        )
        let userPrompt = PromptBuilder.buildUserPrompt(thought: thought, mood: mood, hasAkathisia: hasAkathisia)

        let body: [String: Any] = [
            "systemInstruction": [
                "parts": [["text": systemPrompt]],
            ],
            "contents": [
                [
                    "role": "user",
                    "parts": [["text": userPrompt]],
                ],
            ],
            "generationConfig": [
                "temperature": 0.7,
                "maxOutputTokens": strategy == .crisis ? 512 : 1024,
            ] as [String: Any],
        ]

        let (data, response) = try await performGenerateContent(modelId: model.id, apiKey: apiKey, body: body)
        guard let httpResponse = response as? HTTPURLResponse else {
            throw AIServiceError.invalidResponse
        }

        switch httpResponse.statusCode {
        case 200: break
        case 401, 403: throw AIServiceError.invalidKey
        case 429: throw AIServiceError.httpStatus(429)
        case 500, 502, 503, 504: throw AIServiceError.httpStatus(httpResponse.statusCode)
        default:
            if let msg = geminiErrorMessage(from: data) {
                print("[CBTReframe][Gemini] HTTP \(httpResponse.statusCode): \(msg)")
            }
            throw AIServiceError.httpStatus(httpResponse.statusCode)
        }

        return try parseGeminiGenerateResponse(data, strategy: strategy)
    }

    func analyzeThoughtPatterns(
        thoughts: [ThoughtEntry],
        model: AIModel
    ) async throws -> ThoughtPatternReport {
        guard let apiKey = KeychainManager.shared.load(key: provider.rawValue),
              !apiKey.isEmpty else {
            throw AIServiceError.noAPIKey
        }

        let body: [String: Any] = [
            "systemInstruction": [
                "parts": [["text": PromptBuilder.thoughtPatternSystemPrompt]],
            ],
            "contents": [
                [
                    "role": "user",
                    "parts": [["text": PromptBuilder.buildThoughtPatternUserPrompt(thoughts: thoughts)]],
                ],
            ],
            "generationConfig": [
                "temperature": 0.3,
                "maxOutputTokens": 1400,
            ] as [String: Any],
        ]

        let (data, response) = try await performGenerateContent(modelId: model.id, apiKey: apiKey, body: body)
        guard let httpResponse = response as? HTTPURLResponse else {
            throw AIServiceError.invalidResponse
        }

        switch httpResponse.statusCode {
        case 200: break
        case 401, 403: throw AIServiceError.invalidKey
        case 429: throw AIServiceError.httpStatus(429)
        case 500, 502, 503, 504: throw AIServiceError.httpStatus(httpResponse.statusCode)
        default: throw AIServiceError.httpStatus(httpResponse.statusCode)
        }

        let text = try extractGeminiText(from: data)
        return try parseThoughtPatternContent(text)
    }

    private func performGenerateContent(modelId: String, apiKey: String, body: [String: Any]) async throws -> (Data, URLResponse) {
        guard var components = URLComponents(string: "\(provider.baseURL)/models/\(modelId):generateContent") else {
            throw AIServiceError.invalidResponse
        }
        components.queryItems = [URLQueryItem(name: "key", value: apiKey)]
        guard let url = components.url else {
            throw AIServiceError.invalidResponse
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        // 单次请求上限（与 session 的 request/resource 超时配合）；Flash 通常较快，Pro / 高峰仍可能久等首包
        request.timeoutInterval = 300

        return try await Self.generateSession.data(for: request)
    }

    private func geminiErrorMessage(from data: Data) -> String? {
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let err = json["error"] as? [String: Any] else { return nil }
        return err["message"] as? String ?? err["status"] as? String
    }

    private func parseGeminiGenerateResponse(_ data: Data, strategy: ResponseStrategy) throws -> AnalysisResult {
        let text = try extractGeminiText(from: data)
        let sanitized = LLMJSONSanitizer.sanitizeForJSONObject(text)
        return try parseReframeOutput(sanitized, strategy: strategy)
    }

    private func extractGeminiText(from data: Data) throws -> String {
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw AIServiceError.invalidResponse
        }
        if let err = json["error"] as? [String: Any], let msg = err["message"] as? String {
            throw AIServiceError.parseError(msg)
        }
        guard let candidates = json["candidates"] as? [[String: Any]],
              let first = candidates.first,
              let content = first["content"] as? [String: Any],
              let parts = content["parts"] as? [[String: Any]],
              let text = parts.first?["text"] as? String,
              !text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            throw AIServiceError.invalidResponse
        }
        return text
    }
}

File: shared/src/commonMain/kotlin/com/henryliu/cbtreframe/shared/GeminiService.kt
package com.henryliu.cbtreframe.shared

import io.ktor.client.HttpClient
import io.ktor.client.request.header
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.client.statement.bodyAsText
import io.ktor.http.ContentType
import io.ktor.http.contentType
import kotlinx.serialization.Serializable
import kotlinx.serialization.SerialName
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.contentOrNull

@Serializable
data class GeminiGenerateBody(
    @SerialName("systemInstruction")
    val systemInstruction: GeminiSystemInstruction? = null,
    val contents: List<GeminiContent>,
    @SerialName("generationConfig")
    val generationConfig: GeminiGenerationConfig? = null,
)

@Serializable
data class GeminiSystemInstruction(
    val parts: List<GeminiPart>,
)

@Serializable
data class GeminiContent(
    val role: String? = null,
    val parts: List<GeminiPart>,
)

@Serializable
data class GeminiPart(
    val text: String,
)

@Serializable
data class GeminiGenerationConfig(
    val temperature: Double = 0.7,
    @SerialName("maxOutputTokens")
    val maxOutputTokens: Int = 1024,
)

@Serializable
data class GeminiGenerateResponse(
    val candidates: List<GeminiCandidate> = emptyList(),
    val error: GeminiError? = null,
)

@Serializable
data class GeminiCandidate(
    val content: GeminiContent? = null,
)

@Serializable
data class GeminiError(
    val message: String? = null,
    val status: String? = null,
)

class GeminiService(
    private val httpClient: HttpClient,
    private val apiKeyProvider: suspend () -> String?,
) : AIServiceProtocol {
    override val provider: AIProvider = AIProvider.GEMINI
    private val baseUrl = "https://generativelanguage.googleapis.com/v1beta"

    override suspend fun reframe(
        thought: String,
        mood: String,
        hasAkathisia: Boolean,
        model: AIModel,
        depth: ThinkingTemplate.AnalysisDepth,
        style: ThinkingTemplate.AppResponseStyle,
        template: ThinkingTemplate,
        strategy: ResponseStrategy,
    ): AnalysisResult {
        val apiKey = apiKeyProvider() ?: throw AIServiceError.NoAPIKey()

        val systemPrompt = PromptBuilder.buildSystemPrompt(
            template = template,
            strategy = strategy,
            depth = depth,
            style = style,
            mood = mood,
            hasAkathisia = hasAkathisia,
        )
        val userPrompt = PromptBuilder.buildUserPrompt(thought, mood, hasAkathisia)

        val body = GeminiGenerateBody(
            systemInstruction = GeminiSystemInstruction(
                parts = listOf(GeminiPart(text = systemPrompt)),
            ),
            contents = listOf(
                GeminiContent(
                    role = "user",
                    parts = listOf(GeminiPart(text = userPrompt)),
                ),
            ),
            generationConfig = GeminiGenerationConfig(
                temperature = 0.7,
                maxOutputTokens = if (strategy == ResponseStrategy.crisis) 512 else 4096,
            ),
        )

        val (data, statusCode) = performGenerateContent(model.modelName, apiKey, body)

        when (statusCode) {
            200 -> { /* ok */ }
            401, 403 -> throw AIServiceError.InvalidKey()
            429 -> throw AIServiceError.HttpStatus(429)
            in 500..599 -> throw AIServiceError.HttpStatus(statusCode)
            else -> {
                val errMsg = geminiErrorMessage(data)
                if (errMsg != null) {
                    println("Gemini HTTP $statusCode: $errMsg")
                }
                throw AIServiceError.HttpStatus(statusCode)
            }
        }

        return parseGeminiGenerateResponse(data, strategy)
    }

    override suspend fun analyzeThoughtPatterns(
        thoughts: List<ThoughtEntry>,
        model: AIModel,
    ): ThoughtPatternReport {
        val apiKey = apiKeyProvider() ?: throw AIServiceError.NoAPIKey()

        val body = GeminiGenerateBody(
            systemInstruction = GeminiSystemInstruction(
                parts = listOf(GeminiPart(text = PromptBuilder.thoughtPatternSystemPrompt)),
            ),
            contents = listOf(
                GeminiContent(
                    role = "user",
                    parts = listOf(GeminiPart(text = PromptBuilder.buildThoughtPatternUserPrompt(thoughts))),
                ),
            ),
            generationConfig = GeminiGenerationConfig(
                temperature = 0.3,
                maxOutputTokens = 1400,
            ),
        )

        val (data, statusCode) = performGenerateContent(model.modelName, apiKey, body)

        when (statusCode) {
            200 -> { /* ok */ }
            401, 403 -> throw AIServiceError.InvalidKey()
            429 -> throw AIServiceError.HttpStatus(429)
            in 500..599 -> throw AIServiceError.HttpStatus(statusCode)
            else -> throw AIServiceError.HttpStatus(statusCode)
        }

        val text = extractGeminiText(data)
        return parseThoughtPatternContent(text)
    }

    private suspend fun performGenerateContent(
        modelId: String,
        apiKey: String,
        body: GeminiGenerateBody,
    ): Pair<String, Int> {
        val url = "$baseUrl/models/$modelId:generateContent?key=$apiKey"

        val response = httpClient.post(url) {
            contentType(ContentType.Application.Json)
            setBody(geminiJson.encodeToString(GeminiGenerateBody.serializer(), body))
        }

        return response.bodyAsText() to response.status.value
    }

    private fun geminiErrorMessage(data: String): String? {
        return try {
            val resp = geminiJson.decodeFromString(GeminiGenerateResponse.serializer(), data)
            resp.error?.message ?: resp.error?.status
        } catch (_: Exception) {
            null
        }
    }

    private fun parseGeminiGenerateResponse(data: String, strategy: ResponseStrategy): AnalysisResult {
        val text = extractGeminiText(data)
        return parseReframeOutput(text, strategy)
    }

    private fun extractGeminiText(data: String): String {
        val jsonObj = try {
            geminiJson.parseToJsonElement(data).jsonObject
        } catch (_: Exception) {
            throw AIServiceError.InvalidResponse()
        }

        val error = jsonObj["error"]?.jsonObject
        if (error != null) {
            val msg = error["message"]?.jsonPrimitive?.contentOrNull ?: "unknown"
            throw AIServiceError.ParseError(msg)
        }

        val candidates = jsonObj["candidates"]?.jsonArray
        val firstCandidate = candidates?.firstOrNull()?.jsonObject
            ?: throw AIServiceError.InvalidResponse()
        val content = firstCandidate["content"]?.jsonObject
            ?: throw AIServiceError.InvalidResponse()
        val parts = content["parts"]?.jsonArray
            ?: throw AIServiceError.InvalidResponse()

        val text = parts.mapNotNull { part ->
            part.jsonObject["text"]?.jsonPrimitive?.contentOrNull
        }.joinToString("")

        if (text.isBlank()) {
            throw AIServiceError.InvalidResponse()
        }

        return text
    }
}

private val geminiJson = Json { ignoreUnknownKeys = true; isLenient = true }
